### PR TITLE
ACL Advanced Permission Reports direct links

### PR DESCRIPTION
**Pull request by brianteeman**
**_This is very much a work in progress and I am posting it here so that better coders can help me improve upon it and also to hopefully make it work reliably._**

## Background
For 7 years Joomla has had an option to see fine grain acl permission reports when you enable Debug in the global configuration. As this is incredibly useful and so many people never even knew it existed this has now been enabled by default in Joomla 3.7
screenshotr20-20-44

screenshotr20-14-59

## Aims
Now that we can see these fine grained acl details wouldnt it be useful to be able to go directly to a specific permission from this page and make any changes. I think it would be so I started to look at the code.

## Status
So far I have only looked at the advanced permissions report for groups. If this works well it can be expanded to users as well.
I have analysed the type of link required to go directly to the permissions tab and with the help of some jquery from @dgt41 that he wrote for another joomla module I believe I have got the basics done.

## Oustanding issues
The links do not always work. I think this is related to the way we have the tab state saved in local storage. This is what the jquery script attempts to workaround but it looks like it needs some more work

screenshotr17-44-32

In theory the first edit link above will go directly to

screenshotr20-23-15

### CODE REVIEW

File: administrator/components/com_users/views/debuggroup/tmpl/default.php [hunk 1]
```diff
@@ -15,6 +15,24 @@
 JHtml::_('bootstrap.tooltip');
 JHtml::_('formbehavior.chosen', 'select');
 
+// Add the js from @dgt41
+JHtml::_('jquery.framework');
+JFactory::getDocument()->addScriptDeclaration('
```
Comment (C-Lodder) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: Can we place move this to a separate file in `media/com_users/js/com_users.js`

Comment (brianteeman) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: when it is working then yes that makes sense :)

Comment (dgt41) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: please not `com_user.js` use instead the naming convention:
[admin||site]-[view name]-[layout name].js
so in this case:
`admin-debuggroup-default.js`

Comment (brianteeman) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: whatever name you want i am happy but for now i would prefer to just get it working :)

Comment (dgt41) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: If you are ok to move this to J4 then we can come up with a better solution than the local storage

Comment (brianteeman) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: `/media/com)users/js/admin-debuggroup-default.js
`
??

Comment (brianteeman) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: I really would like to see this in j3 if at all possible. J4 is so far away and we have j3 for at least 30 months

File: administrator/components/com_users/views/debuggroup/tmpl/default.php [hunk 2]
```diff
@@ -85,18 +85,17 @@
 						</td>
 						<td class="nowrap">
 						<?php
-							$parts = explode( '.', $this->escape($item->name));
+							$parts = explode('.', $this->escape($item->name));
 							if(isset($parts[2]) && $parts[1] === 'category') {
```
Comment (C-Lodder) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: spaced after `if`

File: administrator/components/com_users/views/debuggroup/tmpl/default.php [hunk 3]
```diff
@@ -66,7 +84,29 @@
 							<?php echo $this->escape($item->title); ?>
 						</td>
 						<td class="nowrap">
+						<?php
+							$parts = explode('.', $this->escape($item->name));
+							if(isset($parts[2]) && $parts[1] === 'category') {
```
Comment (C-Lodder) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: space after `if`

File: administrator/components/com_users/views/debuggroup/tmpl/default.php [hunk 4]
```diff
@@ -66,7 +84,29 @@
...
+								$url = JRoute::_('index.php?option=com_categories&view=' . $parts[1] . '&layout=edit&id=' . $parts[2] . '&extension=' . $parts[0] . '#permissions-' . $this->group->id);
+							}
+							
+							elseif(isset($parts[2])) {
```
Comment (C-Lodder) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: space after `elseif`

File: administrator/components/com_users/views/debuggroup/tmpl/default.php [hunk 5]
```diff
@@ -66,7 +84,29 @@
...
+								$url = JRoute::_('index.php?option=' . $parts[0] . '&view=' . $parts[1] . '&layout=edit&id=' . $parts[2] . '#permissions-' . $this->group->id);
+							}
+
+							if(!isset($parts[2]))
```
Comment (C-Lodder) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: space after `if`

File: administrator/components/com_users/views/debuggroup/tmpl/default.php [hunk 6]
```diff
@@ -66,7 +84,29 @@
...
+							elseif(isset($parts[2])) {
+								$url = JRoute::_('index.php?option=' . $parts[0] . '&view=' . $parts[1] . '&layout=edit&id=' . $parts[2] . '#permissions-' . $this->group->id);
+							}
+
+							if(!isset($parts[2])) 
+							$url = JRoute::_('index.php?option=com_config&view=component&component=' . $parts[0] . '#permission-' . $this->group->id);
+							?>
+						
+							<span>
+								<a class="js-permission btn btn-small"  href="<?php echo $url;?>">
```
Comment (C-Lodder) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: remove double space before `href`

File: administrator/components/com_users/views/debuggroup/tmpl/default.php (hunk 2 above)
Comment (brianteeman) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: i can do that but why doesnt the sniff pick that up. is it actually our standard

There are 2259 results in 58 files  of **if(**
2864 results in 301 files of **if (**

Comment (C-Lodder) [inline administrator/components/com_users/views/debuggroup/tmpl/default.php]: not sure, perhaps this hasn't been added....@mbabker ?

**Comment by brianteeman** [comment]
Putting this on hold for now until #14959 is fixed as that bug is giving completely incorrect data to this view making any further use of this view useless (or misleading)

**Comment by brianteeman** [comment]
Removed JS in the hope that when #15748 is merged it wont be needed

**Comment by dgt41** [comment]
I don't think that #15748 will open tabs if an anchor is present in the url (unless I'm missing something there)

**Comment by PhilETaylor** [comment]
> I don't think that #15748 will open tabs if an anchor is present in the url (unless I'm missing something there)

Not yet - the code is written that will grap an anchor from the loading url, use that instead of the tab state to click a tab, and then remove the #anchor from the url - it works locally but has not been pushed, I need sleep before I do that and today drained me.

**Comment by brianteeman** [comment]
closing until i have time to work on it

**Pull request closed without merging**